---
title: Analyzing Your Bulk RNA-seq Assay Data
---

## Step 1: Assess the Quality of Your Sequencing Data with [`FastQC`](/pages/packages/fastqc.qmd)
If you do not already have FastQC installed, please refer to [**Installing FastQC**](/pages/installation/fastqc_install.qmd).

You can launch the FastQC graphical interface by simply running:

```bash
fastqc
```

Alternatively, you can run FastQC directly from the command line:

```bash
fastqc --extract --outdir=outdir somefile.fastq someotherfile.fastq ...
```

where you list all the FASTQ files you wish to analyze after the `fastqc` command, and `outdir` is the directory in which you would like the output files to be stored.

For a detailed tutorial on using FastQC and interpreting the output reports, see our page on [**Quality Control of Raw Sequencing Data with FastQC**](/pages/packages/fastqc.qmd).

## Step 2: Pseudoalign Your Library Reads with [`kb-python`](/pages/packages/kb_python.qmd)

Aligning bulk RNA-seq reads is relatively straightforward. After generating a reference index with `kb ref`, pseudoalign your reads to the reference with `kb count`.

### An Example
Here we provide a brief example of how to pseudoalign a bulk library to a reference using `kb count`. For a more in-depth tutorial, see our page on [**kb-python**](/pages/packages/kb_python.qmd).

To psuedoalign your bulk data, run:

```bash
kb count --h5ad -x BULK -g t2g.txt -i index.idx \
  R1.fastq.gz R2.fastq.gz
```

where index.idx and t2g.txt are files generated by `kb ref`, and R1.fastq.gz and R2.fastq.gz are your raw paired-end reads. The `--h5ad` argument outputs an `.h5ad` file suitable for downstream processing with [**Scanpy**](https://scanpy.readthedocs.io/en/stable/).

If you performed a **bulk nucleus RNA-seq** experiment, generate your index with ``kb ref`` using the **NAC workflow**. Then, run `kb count` with the standard workflow as shown above.

## Step 3: Process and Analyze Your Data
Now you are ready to process your data! To continue with downstream analysis, follow our [**example bulk rna-seq pipeline**](/notebooks/bulk_pipeline.ipynb).