<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Quality Control of Raw Sequencing Data with FastQC – CBRC-SPEC Tutorials</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-1df33b310b25e824b5acc70bcebddb02.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d6053df382244508c1004bc951ec3d33.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
<link rel="stylesheet" href="fastqc_figures.css">
</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">CBRC-SPEC Tutorials</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-technologies" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Technologies</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-technologies">    
        <li>
    <a class="dropdown-item" href="./10x.html">
 <span class="dropdown-text">10x Genomics Chromium</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./parse.html">
 <span class="dropdown-text">Parse Evercode WT</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./seqfish_pipeline.html">
 <span class="dropdown-text">seqFISH</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-seqspec" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">seqspec</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-seqspec">    
        <li>
    <a class="dropdown-item" href="./seqspec.html">
 <span class="dropdown-text">What is seqspec?</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./seqspec_tutorials.html">
 <span class="dropdown-text">Tutorials</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-packages" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Packages</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-packages">    
        <li>
    <a class="dropdown-item" href="./fastqc.html">
 <span class="dropdown-text">FastQC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./kb_python.html">
 <span class="dropdown-text">kb-python</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-workflows" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Workflows</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-workflows">    
        <li>
    <a class="dropdown-item" href="./sample_pipeline.html">
 <span class="dropdown-text">An Example scRNA-seq Workflow</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="./seqfish_pipeline.html">
 <span class="dropdown-text">An Example Spatial Transcriptomics Workflow</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-installation" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Installation</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-installation">    
        <li>
    <a class="dropdown-item" href="./conda.html">
 <span class="dropdown-text">Creating a Conda Environment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./python_packages.html">
 <span class="dropdown-text">Installing Python Packages</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./R_packages.html">
 <span class="dropdown-text">Installing R Packages</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fastqc_install.html">
 <span class="dropdown-text">Installing FastQC</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./contact_us.html"> 
<span class="menu-text">Contact Us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#run-fastqc" id="toc-run-fastqc" class="nav-link active" data-scroll-target="#run-fastqc">Run <code>FastQC</code></a></li>
  <li><a href="#a-brief-tutorial-on-fastq-files" id="toc-a-brief-tutorial-on-fastq-files" class="nav-link" data-scroll-target="#a-brief-tutorial-on-fastq-files">A Brief Tutorial on FASTQ Files</a>
  <ul class="collapse">
  <li><a href="#what-are-quality-scores" id="toc-what-are-quality-scores" class="nav-link" data-scroll-target="#what-are-quality-scores">What Are Quality Scores?</a></li>
  </ul></li>
  <li><a href="#understanding-fastqc-plots" id="toc-understanding-fastqc-plots" class="nav-link" data-scroll-target="#understanding-fastqc-plots">Understanding <code>FASTQC</code> Plots</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Quality Control of Raw Sequencing Data with <code>FastQC</code></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><strong>FastQC</strong> is a quality control tool for high-throughput sequencing data. It provides a quick, visual summary of your raw <strong>FASTQ</strong> files, helping you identify potential issues such as low-quality reads, adapter contamination, or biased base composition. FastQC generates easy-to-read HTML reports that summarize key quality metrics like per-base quality scores, GC content, and sequence duplication levels. Running FastQC before alignment or downstream analysis ensures your sequencing data are high-quality and suitable for accurate biological interpretation.</p>
<p>Here we provide a brief tutorial for analyzing your FASTQ files with FastQC. For a more in-depth explanation of FastQC please refer to the <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">website</a> and <a href="https://github.com/s-andrews/FastQC">GitHub page</a>.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Installing FastQC">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Installing FastQC
</div>
</div>
<div class="callout-body-container callout-body">
<p>To get instructions on how to install the FastQC package, see our page on <a href="./fastqc_install.html">Installing FastQC</a>.</p>
</div>
</div>
<section id="run-fastqc" class="level2">
<h2 class="anchored" data-anchor-id="run-fastqc">Run <code>FastQC</code></h2>
<p>FastQC conveniently has a built-in GUI in which to run your analysis. To open the GUI simply type into your command-line:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>and follow the displayed instructions.</p>
<p>If you would prefer to analyze your data through the command-line, simply specify a list of files following the <code>fastqc</code> command:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--extract</span> <span class="at">--outdir</span><span class="op">=&lt;</span>outdir<span class="op">&gt;</span> somefile.fastq someotherfile.fastq ...</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>and replace &lt;outdir&gt; with the path to the directory you wish to have your output stored.</p>
</section>
<section id="a-brief-tutorial-on-fastq-files" class="level2">
<h2 class="anchored" data-anchor-id="a-brief-tutorial-on-fastq-files">A Brief Tutorial on FASTQ Files</h2>
<p>To understand the FastQC output, we must first understand what a FASTQ file is and how it is generated.</p>
<p>FASTQ files are the standard format for storing raw sequencing reads from next-generation sequencing instruments. Each read in a FASTQ file is represented by four lines:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode ruby code-with-copy"><code class="sourceCode ruby"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ot">@SEQ_ID</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="cn">GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="op">+</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span><span class="vs">''</span><span class="op">*</span>((((<span class="op">***+</span>))<span class="op">%</span><span class="ot">%%</span><span class="st">++)(</span><span class="ot">%</span><span class="op">%</span><span class="ot">%%</span><span class="st">).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li><p><strong>Line 1</strong> – Read identifier (begins with @)</p></li>
<li><p><strong>Line 2</strong> – Nucleotide sequence</p></li>
<li><p><strong>Line 3</strong> – Separator line (usually just +)</p></li>
<li><p><strong>Line 4</strong> – Quality scores for each base (encoded as ASCII characters)</p></li>
</ul>
<section id="what-are-quality-scores" class="level3">
<h3 class="anchored" data-anchor-id="what-are-quality-scores">What Are Quality Scores?</h3>
<p>Each entry in a FASTQ file contains not only the DNA or RNA sequence itself, but also a <strong>quality score</strong> for every base. These scores reflect the sequencing machine’s confidence in each base call — that is, how likely it is that a given nucleotide (A, T, G, or C) was correctly identified.</p>
<p>FASTQ quality scores are typically encoded using the <strong>Phred scale</strong>, a logarithmic transformation of the error probability:</p>
<p><span class="math display">\[Q = -10\times \log_{10}(P)\]</span></p>
<p>where <span class="math inline">\(P\)</span> is the probability that a base call is incorrect. For example, a Phred score of 20 means there’s a 1% chance the base is wrong, while a score of 30 corresponds to a 0.1% error rate.</p>
<p>These scores are stored as ASCII characters in the FASTQ file, allowing both sequence and quality information to be represented compactly.</p>
<p>During quality control, tools like <strong>FastQC</strong> visualizes these scores across all reads helping identify low-quality regions that may need trimming or filtering before alignment and downstream analysis.</p>
</section>
</section>
<section id="understanding-fastqc-plots" class="level2">
<h2 class="anchored" data-anchor-id="understanding-fastqc-plots">Understanding <code>FASTQC</code> Plots</h2>
<p>Below are the plots that <strong>FastQC</strong> generates, each providing diagnostic insights into your sequencing data quality.</p>
<hr>
<details>
<summary>
<strong>1. Per-Base Sequence Quality Over All Reads</strong>
</summary>
<p>This plot shows the distribution of quality scores for each base position across all reads. Typically, quality is highest near the 5′ end and drops toward the 3′ end of reads.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/per_base_SQ.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>A sharp decline in quality may indicate sequencing problems or the need for trimming low-quality bases before alignment.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/bad/per_base_SQ.png" class="img-fluid figure-img"></p>
<figcaption>A Bad FastQC Report</figcaption>
</figure>
</div>
</details>
<details>
<summary>
<strong>2. Per-Tile Sequence Quality Over All Reads</strong>
</summary>
<p>This graph visualizes quality variation across different tiles of the sequencing flow cell.<br>
Each tile corresponds to a physical area on the sequencer’s imaging surface.</p>
<p>Systematic differences between tiles may point to instrument issues, uneven illumination, or problems during image analysis.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/bad/per_tile_SQ.png" class="img-fluid figure-img"></p>
<figcaption>A Bad FastQC Report</figcaption>
</figure>
</div>
</details>
<details>
<summary>
<strong>3. Quality Score Distribution Over All Sequences</strong>
</summary>
<p>This plot summarizes the overall distribution of mean quality scores per read.<br>
Ideally, most reads should have consistently high average scores.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/per_sequence_QC.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>A broad or bimodal distribution suggests variable read quality, possibly due to uneven sequencing performance or mixed-quality samples.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/bad/per_sequence_QC.png" class="img-fluid figure-img"></p>
<figcaption>A Bad FastQC Report</figcaption>
</figure>
</div>
</details>
<details>
<summary>
<strong>4. Sequence Content Across All Bases</strong>
</summary>
<p>This plot shows the proportion of each nucleotide (A, T, G, and C) at every base position.<br>
In random libraries, the lines should be roughly parallel.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/per_base_SC.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>Deviations early in the read may indicate biased priming, adapter contamination, or non-random base composition typical of certain assays (e.g., single-cell barcoding regions).</p>
</details>
<details>
<summary>
<strong>5. Per Sequence GC Content</strong>
</summary>
<p>This graph compares the GC content distribution of all reads to a theoretical normal distribution.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/per_sequence_GCC.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>A shifted or multi-peaked curve may reflect contamination from other organisms, amplification bias, or targeted sequencing regions with atypical GC content.</p>
</details>
<details>
<summary>
<strong>6. Per Base N Content</strong>
</summary>
<p>This plot shows the percentage of ambiguous bases (“N”) called at each position in the read.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/per_base_NC.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>High or position-specific N content indicates uncertainty in base calling and may suggest low signal intensity or sequencing chemistry problems.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/bad/per_base_NC.png" class="img-fluid figure-img"></p>
<figcaption>A Bad FastQC Report</figcaption>
</figure>
</div>
</details>
<details>
<summary>
<strong>7. Sequence Length Distribution</strong>
</summary>
<p>This plot displays the distribution of read lengths in the dataset.<br>
For most sequencing platforms, a single consistent read length is expected.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/seq_length_dist.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>Variable lengths may indicate trimming, read merging, or incomplete read cycles.</p>
</details>
<details>
<summary>
<strong>8. Sequence Duplication Levels</strong>
</summary>
<p>This plot shows how many sequences are duplicated and how often.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/seq_duplication.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
<p>High duplication levels can indicate PCR amplification bias or low library complexity.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/bad/seq_duplication.png" class="img-fluid figure-img"></p>
<figcaption>A Bad FastQC Report</figcaption>
</figure>
</div>
<p>However, in RNA-seq or scRNA-seq data, some duplication is expected due to highly expressed genes.</p>
</details>
<details>
<summary>
<strong>9. Overrepresented Sequences</strong>
</summary>
<p>This section lists any sequences that occur more frequently than expected by chance.<br>
Overrepresentation may arise from adapter contamination, primer dimers, or abundant biological sequences such as rRNA.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/bad/overrepresented_sequences.png" class="img-fluid figure-img"></p>
<figcaption>A Bad FastQC Report</figcaption>
</figure>
</div>
<p>Identifying and removing these sequences improves downstream analysis accuracy.</p>
</details>
<details>
<summary>
<strong>10. Adapter Content</strong>
</summary>
<p>This plot quantifies how much adapter sequence remains untrimmed at each read position.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastqc_plots/good/adapter_content.png" class="img-fluid figure-img"></p>
<figcaption>A Good FastQC Report</figcaption>
</figure>
</div>
Increasing adapter content toward the 3′ end of reads is a common sign that reads are longer than the actual insert fragment and require adapter trimming before alignment.
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><img src="black_Caltech_logo.png" alt="Caltech Logo" width="300"></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>