<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>sample_pipeline – CBRC-SPEC Tutorials</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-1df33b310b25e824b5acc70bcebddb02.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d6053df382244508c1004bc951ec3d33.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">CBRC-SPEC Tutorials</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-tutorials" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Tutorials</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-tutorials">    
        <li>
    <a class="dropdown-item" href="./10x.html">
 <span class="dropdown-text">10x Genomics Chromium</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./parse.html">
 <span class="dropdown-text">Parse Evercode WT</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./seqfish.html">
 <span class="dropdown-text">seqFISH</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-seqspec" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">seqspec</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-seqspec">    
        <li>
    <a class="dropdown-item" href="./seqspec.html">
 <span class="dropdown-text">What is seqspec?</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./seqspec_tutorials.html">
 <span class="dropdown-text">Tutorials</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./kb_python.html"> 
<span class="menu-text">kb-python</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-workflows" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Workflows</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-workflows">    
        <li>
    <a class="dropdown-item" href="./sample_pipeline.html">
 <span class="dropdown-text">An Example scRNA-seq Pipeline</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="./seqfish.html">
 <span class="dropdown-text">An Example Spatial Transcriptomics Pipeline</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#an-introductory-analysis-pipeline" id="toc-an-introductory-analysis-pipeline" class="nav-link active" data-scroll-target="#an-introductory-analysis-pipeline">An Introductory Analysis Pipeline</a></li>
  <li><a href="#initial-set-up" id="toc-initial-set-up" class="nav-link" data-scroll-target="#initial-set-up">Initial Set-Up</a>
  <ul class="collapse">
  <li><a href="#install-python-packages" id="toc-install-python-packages" class="nav-link" data-scroll-target="#install-python-packages">Install Python Packages</a></li>
  <li><a href="#download-the-scrna-seq-data" id="toc-download-the-scrna-seq-data" class="nav-link" data-scroll-target="#download-the-scrna-seq-data">Download the scRNA-seq data</a></li>
  </ul></li>
  <li><a href="#step-1-read-alignment" id="toc-step-1-read-alignment" class="nav-link" data-scroll-target="#step-1-read-alignment">Step 1: Read Alignment</a>
  <ul class="collapse">
  <li><a href="#download-a-pre-built-index-with-kb-ref" id="toc-download-a-pre-built-index-with-kb-ref" class="nav-link" data-scroll-target="#download-a-pre-built-index-with-kb-ref">Download a Pre-Built Index with <code>kb ref</code></a></li>
  <li><a href="#pseudoalign-the-scrna-seq-data-to-the-index-with-kb-count" id="toc-pseudoalign-the-scrna-seq-data-to-the-index-with-kb-count" class="nav-link" data-scroll-target="#pseudoalign-the-scrna-seq-data-to-the-index-with-kb-count">Pseudoalign the scRNA-seq Data to the Index with <code>kb count</code></a></li>
  <li><a href="#load-the-anndata-object" id="toc-load-the-anndata-object" class="nav-link" data-scroll-target="#load-the-anndata-object">Load the Anndata Object</a></li>
  </ul></li>
  <li><a href="#step-2-basic-quality-control" id="toc-step-2-basic-quality-control" class="nav-link" data-scroll-target="#step-2-basic-quality-control">Step 2: Basic Quality Control</a>
  <ul class="collapse">
  <li><a href="#test-for-library-saturation" id="toc-test-for-library-saturation" class="nav-link" data-scroll-target="#test-for-library-saturation">Test for Library Saturation</a></li>
  <li><a href="#examine-the-knee-plot" id="toc-examine-the-knee-plot" class="nav-link" data-scroll-target="#examine-the-knee-plot">Examine the Knee Plot</a></li>
  </ul></li>
  <li><a href="#step-3-doublet-and-multiplet-removal" id="toc-step-3-doublet-and-multiplet-removal" class="nav-link" data-scroll-target="#step-3-doublet-and-multiplet-removal">Step 3: Doublet and Multiplet Removal</a>
  <ul class="collapse">
  <li><a href="#filter-empty-droplets-according-to-the-knee-plot" id="toc-filter-empty-droplets-according-to-the-knee-plot" class="nav-link" data-scroll-target="#filter-empty-droplets-according-to-the-knee-plot">Filter empty droplets (According to the Knee Plot)</a></li>
  <li><a href="#filter-multiplets-with-scrublet" id="toc-filter-multiplets-with-scrublet" class="nav-link" data-scroll-target="#filter-multiplets-with-scrublet">Filter Multiplets with Scrublet</a></li>
  <li><a href="#filtering-cells-by-mitochondrial-content" id="toc-filtering-cells-by-mitochondrial-content" class="nav-link" data-scroll-target="#filtering-cells-by-mitochondrial-content">Filtering Cells by Mitochondrial Content</a></li>
  <li><a href="#filter-out-genes-that-are-not-present-in-any-cells" id="toc-filter-out-genes-that-are-not-present-in-any-cells" class="nav-link" data-scroll-target="#filter-out-genes-that-are-not-present-in-any-cells">Filter out genes that are not present in any cells</a></li>
  <li><a href="#visualizing-count-distributions" id="toc-visualizing-count-distributions" class="nav-link" data-scroll-target="#visualizing-count-distributions">Visualizing count distributions</a></li>
  </ul></li>
  <li><a href="#step-4-normalization-and-variance-stabilization" id="toc-step-4-normalization-and-variance-stabilization" class="nav-link" data-scroll-target="#step-4-normalization-and-variance-stabilization">Step 4: Normalization and Variance Stabilization</a>
  <ul class="collapse">
  <li><a href="#identify-highly-variable-genes" id="toc-identify-highly-variable-genes" class="nav-link" data-scroll-target="#identify-highly-variable-genes">Identify Highly Variable Genes</a></li>
  </ul></li>
  <li><a href="#step-5-clustering-and-visualization" id="toc-step-5-clustering-and-visualization" class="nav-link" data-scroll-target="#step-5-clustering-and-visualization">Step 5: Clustering and Visualization</a>
  <ul class="collapse">
  <li><a href="#visualization-with-pca" id="toc-visualization-with-pca" class="nav-link" data-scroll-target="#visualization-with-pca">Visualization with PCA</a></li>
  <li><a href="#nonlinear-visualization" id="toc-nonlinear-visualization" class="nav-link" data-scroll-target="#nonlinear-visualization">Nonlinear Visualization</a></li>
  </ul></li>
  <li><a href="#step-6-differential-analysis" id="toc-step-6-differential-analysis" class="nav-link" data-scroll-target="#step-6-differential-analysis">Step 6: Differential Analysis</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"></header>




<div class="colab-button-container">
<p><a href="https://colab.research.google.com/github/username/project/blob/main/notebooks/scrna_tutorial.ipynb" target="_blank" class="colab-top-button" title="Open this notebook in Google Colab"> <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" class="colab-badge"> </a></p>
</div>
<h1>
Introductory scRNA-seq Analysis with <code>kb-python</code> and <code>ScanPy</code>
</h1>
<h4 class="anchored">
Written by Maya Caskey (based off older versions by Joe Rich, Kyung Hoi (Joseph) Min, A. Sina Booeshaghi and Lior Pachter)
</h4>
<p><br></p>
<p>This notebook demonstrates pre-processing and basic analysis of the <a href="https://www.10xgenomics.com/datasets/1-k-human-pbm-cs-stained-with-a-panel-of-total-seq-b-antibodies-single-indexed-3-1-standard-4-0-0">1k Human PBMCs</a> dataset from 10x Genomics using the <strong>10x Genomics Chromium Single Cell 3’ Next Gen v3.1</strong> assay.</p>
<section id="an-introductory-analysis-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="an-introductory-analysis-pipeline">An Introductory Analysis Pipeline</h2>
<p>There are many ways to perform a single-cell RNA sequencing (scRNA-seq) analysis and many aspects of your data to explore. In this tutorial, we outline a basic approach to quality control and analysis, along with some intuitive explanation of the underlying mathematics.</p>
<div class="highlight-section">
<p>This scRNA-seq workflow has the following steps:</p>
<ol type="1">
<li>Read Alignment</li>
<li>Quality Control</li>
<li>Cell and Gene Filters</li>
<li>Normalization and Variance Stabilization</li>
<li>Dimensionality Reduction</li>
<li>Clustering and Visualization</li>
<li>Differential Expression</li>
</ol>
</div>
</section>
<section id="initial-set-up" class="level2">
<h2 class="anchored" data-anchor-id="initial-set-up">Initial Set-Up</h2>
<section id="install-python-packages" class="level3">
<h3 class="anchored" data-anchor-id="install-python-packages">Install Python Packages</h3>
<div id="cell-8" class="cell" data-cellview="form">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># @title</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">--</span>quiet scanpy python<span class="op">-</span>igraph louvain pybiomart</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">--</span>quiet matplotlib</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">--</span>quiet scikit<span class="op">-</span>learn</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">--</span>quiet numpy</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">--</span>quiet scipy</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install <span class="op">--</span>quiet kb<span class="op">-</span>python<span class="op">==</span><span class="fl">0.29.1</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="cell-9" class="cell" data-cellview="form" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="cb922f35-c2ee-4252-d002-5d9510ae3db8">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># @title</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Import packages</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib_inline.backend_inline <span class="im">import</span> set_matplotlib_formats</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> TruncatedSVD</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> sparse, io</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>matplotlib.rcParams.update({<span class="st">'font.size'</span>: <span class="dv">12</span>})</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>set_matplotlib_formats(<span class="st">'retina'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/_utils/__init__.py:33: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.
  from anndata import __version__ as anndata_version
/usr/local/lib/python3.12/dist-packages/scanpy/__init__.py:24: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.
  if Version(anndata.__version__) &gt;= Version("0.11.0rc2"):
/usr/local/lib/python3.12/dist-packages/scanpy/readwrite.py:16: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.
  if Version(anndata.__version__) &gt;= Version("0.11.0rc2"):</code></pre>
</div>
</div>
</section>
<section id="download-the-scrna-seq-data" class="level3">
<h3 class="anchored" data-anchor-id="download-the-scrna-seq-data">Download the scRNA-seq data</h3>
<div id="cell-11" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>wget <span class="op">-</span>q https:<span class="op">//</span>cf<span class="fl">.10</span><span class="er">xgenomics</span>.com<span class="op">/</span>samples<span class="op">/</span>cell<span class="op">-</span>exp<span class="op">/</span><span class="fl">4.0.0</span><span class="op">/</span>SC3_v3_NextGem_SI_PBMC_CSP_1K<span class="op">/</span>SC3_v3_NextGem_SI_PBMC_CSP_1K_fastqs.tar</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>tar <span class="op">-</span>xf SC3_v3_NextGem_SI_PBMC_CSP_1K_fastqs.tar</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="step-1-read-alignment" class="level2">
<h2 class="anchored" data-anchor-id="step-1-read-alignment">Step 1: Read Alignment</h2>
<section id="download-a-pre-built-index-with-kb-ref" class="level3">
<h3 class="anchored" data-anchor-id="download-a-pre-built-index-with-kb-ref">Download a Pre-Built Index with <code>kb ref</code></h3>
<div id="cell-14" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="d7f50e9c-6a87-475f-ba0b-ac624deb60b0">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>kb ref <span class="op">-</span>d human <span class="op">-</span>i index.idx <span class="op">-</span>g t2g.txt</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[2025-10-21 04:48:16,146]    INFO [download] Skipping download because some files already exist. Use the --overwrite flag to overwrite.</code></pre>
</div>
</div>
</section>
<section id="pseudoalign-the-scrna-seq-data-to-the-index-with-kb-count" class="level3">
<h3 class="anchored" data-anchor-id="pseudoalign-the-scrna-seq-data-to-the-index-with-kb-count">Pseudoalign the scRNA-seq Data to the Index with <code>kb count</code></h3>
<div id="cell-16" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="aedcf646-bef1-457f-d8ef-dc4570369240">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This step runs `kb` to pseudoalign the reads, and then generate the cells x gene matrix in h5ad format.</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>kb count <span class="op">-</span>i index.idx <span class="op">-</span>g t2g.txt <span class="op">-</span>x <span class="dv">10</span><span class="er">XV3</span> <span class="op">--</span>h5ad <span class="op">-</span>t <span class="dv">2</span> <span class="op">\</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_S1_L002_R1_001.fastq.gz <span class="op">\</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_S1_L002_R2_001.fastq.gz <span class="op">\</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_S1_L003_R1_001.fastq.gz <span class="op">\</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_S1_L003_R2_001.fastq.gz <span class="op">\</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_S1_L004_R1_001.fastq.gz <span class="op">\</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_fastqs<span class="op">/</span>SC3_v3_NextGem_SI_CSP<span class="op">-</span>Labeled_PBMCs_1K_gex_S1_L004_R2_001.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[2025-10-21 04:48:27,652]    INFO [count] Skipping kallisto bus because output files already exist. Use the --overwrite flag to overwrite.
[2025-10-21 04:48:27,652]    INFO [count] Sorting BUS file ./output.bus to ./tmp/output.s.bus
[2025-10-21 04:48:46,541]    INFO [count] On-list not provided
[2025-10-21 04:48:46,541]    INFO [count] Copying pre-packaged 10XV3 on-list to .
[2025-10-21 04:48:47,999]    INFO [count] Inspecting BUS file ./tmp/output.s.bus
[2025-10-21 04:49:06,258]    INFO [count] Correcting BUS records in ./tmp/output.s.bus to ./tmp/output.s.c.bus with on-list ./10x_version3_whitelist.txt
[2025-10-21 04:49:31,255]    INFO [count] Sorting BUS file ./tmp/output.s.c.bus to ./output.unfiltered.bus
[2025-10-21 04:49:38,691]    INFO [count] Generating count matrix ./counts_unfiltered/cells_x_genes from BUS file ./output.unfiltered.bus
[2025-10-21 04:49:47,233]    INFO [count] Writing gene names to file ./counts_unfiltered/cells_x_genes.genes.names.txt
[2025-10-21 04:49:47,817] WARNING [count] 13914 gene IDs do not have corresponding valid gene names. These genes will use their gene IDs instead.
[2025-10-21 04:49:47,866]    INFO [count] Reading matrix ./counts_unfiltered/cells_x_genes.mtx
[2025-10-21 04:49:48,887]    INFO [count] Writing matrix to h5ad ./counts_unfiltered/adata.h5ad</code></pre>
</div>
</div>
<p>When the <code>--h5ad</code> argument is used, <code>kb count</code> generates a H5AD-formatted Anndata matrix for downstream processing.</p>
</section>
<section id="load-the-anndata-object" class="level3">
<h3 class="anchored" data-anchor-id="load-the-anndata-object">Load the Anndata Object</h3>
<p>For this tutorial we will be using the Python package <a href="https://scanpy.readthedocs.io/en/stable/"><strong>ScanPy</strong></a>.</p>
<p>ScanPy is a scalable Python library for analyzing single-cell RNA sequencing (scRNA-seq) data, built around an efficient <a href="https://anndata.readthedocs.io/en/stable/"><strong>AnnData</strong></a> object that stores expression matrices and associated metadata. It provides a comprehensive toolkit for preprocessing, visualization, clustering, and differential expression, enabling end-to-end analysis of large single-cell datasets.</p>
<p>Below we load the <em>cells x genes</em> matrix generated by kb-python as a H5AD-formatted object so that we can analyze our data with ScanPy.</p>
<div id="cell-19" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="6374510d-60d0-473e-c479-398a6abf8c3c">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import data</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> anndata.read_h5ad(<span class="st">'counts_unfiltered/adata.h5ad'</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>adata</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="61">
<pre><code>AnnData object with n_obs × n_vars = 281262 × 39546</code></pre>
</div>
</div>
<div id="cell-20" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the gene names and make them a column in the var dataframe</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"counts_unfiltered/cells_x_genes.genes.names.txt"</span>, <span class="st">'r'</span>) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  adata.var[<span class="st">'gene_names'</span>] <span class="op">=</span> <span class="bu">file</span>.readlines()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="step-2-basic-quality-control" class="level2">
<h2 class="anchored" data-anchor-id="step-2-basic-quality-control">Step 2: Basic Quality Control</h2>
<section id="test-for-library-saturation" class="level3">
<h3 class="anchored" data-anchor-id="test-for-library-saturation">Test for Library Saturation</h3>
<p>Below is a <strong>Gene vs.&nbsp;UMI plot</strong>, in which, for each cell, we visualize how many genes we detected. The purpose of this plot is to see if we have “saturated” our sequencing library and sampled a representative number of mRNA molecules for each cell. If this is the case, then increasing the number of UMI counts (x-axis) will not yield an appreciable increase in the number of genes detected (y-axis).</p>
<div id="cell-24" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:625}}" data-outputid="2304f482-ed5e-4eeb-ebdc-7ea6b46a396a">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a plot showing genes detected as a function of UMI counts.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">7</span>))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> np.asarray(adata.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>))[:,<span class="dv">0</span>]</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> np.asarray(np.<span class="bu">sum</span>(adata.X<span class="op">&gt;</span><span class="dv">0</span>, axis<span class="op">=</span><span class="dv">1</span>))[:,<span class="dv">0</span>]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>ax.scatter(x, y, color<span class="op">=</span><span class="st">"green"</span>, alpha<span class="op">=</span><span class="fl">0.25</span>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"UMI Counts"</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Genes Detected"</span>)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>ax.set_yscale(<span class="st">'log'</span>, nonpositive<span class="op">=</span><span class="st">'clip'</span>)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>ax.set_xlim((<span class="fl">0.5</span>, <span class="dv">4500</span>))</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>ax.set_ylim((<span class="fl">0.5</span>,<span class="dv">2000</span>))</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-9-output-1.png" width="855" height="608" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The data looks good! The number of genes does not increase appreciably relative to the number of UMI counts as demonstrated by the linear relationship between genes to UMI counts in the above log-log plot.</p>
</section>
<section id="examine-the-knee-plot" class="level3">
<h3 class="anchored" data-anchor-id="examine-the-knee-plot">Examine the Knee Plot</h3>
<p>In <strong>microfluidics-based scRNA-seq assays</strong>, such as the 10x Genomics Chromium platform, cells are encapsulated into droplets at low concentrations to minimize the formation of <strong>multiplets</strong> (droplets containing more than one cell). However, this approach also results in many droplets that contain no cells at all.</p>
<p>During cell dissociation and droplet generation, some mRNA molecules are released into the surrounding solution — known as <strong>ambient RNA</strong>. When this ambient RNA is captured and sequenced, <strong>empty droplets</strong> may falsely appear to contain a cell.</p>
<p>The <strong>knee plot</strong> helps distinguish real cells from empty droplets based on total UMI counts per droplet. It was first introduced in the <a href="https://www.cell.com/fulltext/S0092-8674(15)00549-8">Drop-seq paper by Macosko et al., 2015</a>.</p>
<p>In this plot, barcodes (droplets) are ranked by the number of associated UMI counts (shown on the x-axis), while the fraction of droplets with at least that many UMIs is shown on the y-axis. The “knee” of the curve represents the point where UMI counts drop sharply — separating droplets that likely contain real cells from those containing only ambient RNA.</p>
<div id="cell-27" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:625}}" data-outputid="938d1a5d-0f04-4009-f83e-0a5471123d7f">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#@title Threshold Cells According to Knee Plot { run: "auto", vertical-output: true }</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>cutoff <span class="op">=</span> <span class="dv">400</span>  <span class="co">#@param {type:"integer"}</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>knee <span class="op">=</span> np.sort((np.array(adata.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>))).flatten())[::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>cell_set <span class="op">=</span> np.arange(<span class="bu">len</span>(knee))</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>num_cells <span class="op">=</span> cell_set[knee <span class="op">&gt;</span> cutoff][::<span class="op">-</span><span class="dv">1</span>][<span class="dv">0</span>]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">7</span>))</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>ax.loglog(knee, cell_set, linewidth<span class="op">=</span><span class="dv">5</span>, color<span class="op">=</span><span class="st">"g"</span>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>ax.axvline(x<span class="op">=</span>cutoff, linewidth<span class="op">=</span><span class="dv">3</span>, color<span class="op">=</span><span class="st">"k"</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>ax.axhline(y<span class="op">=</span>num_cells, linewidth<span class="op">=</span><span class="dv">3</span>, color<span class="op">=</span><span class="st">"k"</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"UMI Counts"</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Set of Barcodes"</span>)</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>plt.grid(<span class="va">True</span>, which<span class="op">=</span><span class="st">"both"</span>)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-10-output-1.png" width="855" height="608" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-28" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="0cf3cf6c-975d-4c81-8d6f-b6f6dfa3cfd0">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>num_cells<span class="sc">:,.0f}</span><span class="ss"> cells passed the </span><span class="sc">{</span>cutoff<span class="sc">}</span><span class="ss"> UMI threshold"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>1,212 cells passed the 400 UMI threshold</code></pre>
</div>
</div>
<p>The knee plot can be used to threshold cells based on the number of UMI counts they contain. In this example we use the number 3979 based on the publication describing the data.</p>
</section>
</section>
<section id="step-3-doublet-and-multiplet-removal" class="level2">
<h2 class="anchored" data-anchor-id="step-3-doublet-and-multiplet-removal">Step 3: Doublet and Multiplet Removal</h2>
<section id="filter-empty-droplets-according-to-the-knee-plot" class="level3">
<h3 class="anchored" data-anchor-id="filter-empty-droplets-according-to-the-knee-plot">Filter empty droplets (According to the Knee Plot)</h3>
<p>Ideally, your data will contain no empty droplets that can be mistaken for real cells. We use our results from the knee plot here to filter them out.</p>
<div id="cell-33" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the cells according to the threshold determined from the knee plot</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, min_genes<span class="op">=</span><span class="dv">200</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, min_counts<span class="op">=</span>knee[num_cells])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The knee plot is a very simple method for getting rid of empty droplets. More sophisticated methods also exist. For instance, the methods <strong>EmptyDrops</strong>, <a href="https://github.com/broadinstitute/CellBender"><strong>CellBender</strong></a>, and <a href="https://bioconductor.org/packages/release/bioc/html/decontX.html"><strong>DecontX</strong></a> use a statistical model to distinguish counts from ambient vs.&nbsp;cellular mRNA. This allows for both the identification and removal of empty droplets and the removal of technical noise from your count data.</p>
</div>
</div>
</section>
<section id="filter-multiplets-with-scrublet" class="level3">
<h3 class="anchored" data-anchor-id="filter-multiplets-with-scrublet">Filter Multiplets with Scrublet</h3>
<p><a href="https://github.com/swolock/scrublet"><strong>Scrublet</strong></a> is an package designed to identify and remove multiplets (droplets containing more than one cell) from scRNA-seq data. It works by simulating artificial doublets through random pairwise combinations of real cells’ expression profiles. Scrublet then embeds both the simulated and observed cells in a low-dimensional space and flags real cells that lie close to simulated doublets as potential multiplets.</p>
<p>Scrublet operates on a raw count matrix, but requires that empty droplets be removed beforehand. If empty droplets remain in the dataset, they can distort the simulated doublet distribution and lead to inaccurate classifications. In addition, the Scrublet package should be applied to each sample/batch individually to avoid batch effects distorting the simulated doublet distribution.</p>
<p>Removing multiplets is an important quality-control step, as doublets can appear as hybrid cell types and obscure true biological variation during clustering and downstream analyses.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The Scrublet package is convenient because it is built-in to ScanPy, but it is not the only nor the most accurate doublet removal method. For a review of doublet detection methods, refer to the paper <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC7897250/">Benchmarking computational doublet-detection methods for single-cell RNA sequencing data</a> by Nan Miles Chi and Jingyi Jessica Li.</p>
</div>
</div>
<div id="cell-38" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="12bfbb7e-9df9-4a62-ff67-26aa90b5c7ec">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sc.pp.scrublet(adata, expected_doublet_rate<span class="op">=</span><span class="fl">0.008</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/neighbors/__init__.py:430: FutureWarning: Use obsm (e.g. `k in adata.obsm` or `adata.obsm.keys() | {'u'}`) instead of AnnData.obsm_keys, AnnData.obsm_keys is deprecated and will be removed in the future.
  if "X_diffmap" in adata.obsm_keys():</code></pre>
</div>
</div>
<p>The expected doublet rate depends upon your assay and the number of cells targeted. Refer to the documentation for you specific technology to get this number. We got the expected double rate 0.008 for 1,000 cells for 10x Chromium NextGem v3.1 <a href="https://kb.10xgenomics.com/s/article/360054599512-What-is-the-cell-multiplet-rate-when-using-the-3-CellPlex-Kit-for-Cell-Multiplexing">here</a>.</p>
<div id="cell-40" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs[<span class="st">'predicted_doublet'</span>] <span class="op">==</span> <span class="va">False</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="filtering-cells-by-mitochondrial-content" class="level3">
<h3 class="anchored" data-anchor-id="filtering-cells-by-mitochondrial-content">Filtering Cells by Mitochondrial Content</h3>
<p>A healthy, intact cell should have no miRNA in the cytoplasm. As such, a high percentage of miRNA in the cytosol indicates cellular damage. We will identify the number of miRNA in each cell and remove cells with a certain percentage of miRNA.</p>
<div id="cell-43" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="801f5ec5-c704-4a90-8812-e4fd5bc553f4">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mito_ensembl_ids <span class="op">=</span> sc.queries.mitochondrial_genes(<span class="st">"hsapiens"</span>, attrname<span class="op">=</span><span class="st">"ensembl_gene_id"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="op">=</span> <span class="bu">set</span>(mito_ensembl_ids[<span class="st">"ensembl_gene_id"</span>].values)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>adata_base_var_names <span class="op">=</span> adata.var_names.<span class="bu">str</span>.split(<span class="st">'.'</span>).<span class="bu">str</span>[<span class="dv">0</span>]  <span class="co"># Removes minor version from var names</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>mito_genes_base <span class="op">=</span> {gene.split(<span class="st">'.'</span>)[<span class="dv">0</span>] <span class="cf">for</span> gene <span class="kw">in</span> mito_genes}  <span class="co"># Removes minor version from mito_genes</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify mitochondrial genes in adata.var using the stripped version of gene IDs</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'is_mito'</span>] <span class="op">=</span> adata_base_var_names.isin(mito_genes_base)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>mito_counts <span class="op">=</span> adata[:, adata.var[<span class="st">'is_mito'</span>]].X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate total counts per cell</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>total_counts <span class="op">=</span> adata.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate percent mitochondrial gene expression per cell</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'percent_mito'</span>] <span class="op">=</span> np.array(mito_counts <span class="op">/</span> total_counts <span class="op">*</span> <span class="dv">100</span>).flatten()</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'n_counts'</span>] <span class="op">=</span> adata.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>).A1</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipython-input-2771333037.py:8: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.
  adata.var['is_mito'] = adata_base_var_names.isin(mito_genes_base)</code></pre>
</div>
</div>
<div id="cell-44" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:491}}" data-outputid="3928cefc-ffff-463b-8922-bbfba6f0a9d2">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'n_counts'</span>, y<span class="op">=</span><span class="st">'percent_mito'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/plotting/_anndata.py:397: FutureWarning: Use obs (e.g. `k in adata.obs` or `str(adata.obs.columns.tolist())`) instead of AnnData.obs_keys, AnnData.obs_keys is deprecated and will be removed in the future.
  if key in adata.obs_keys():</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-16-output-2.png" width="615" height="418" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We see above that the highest density of cells appear to have less than 30% miRNA content, so we will set our threshold to 30%</p>
<div id="cell-46" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs.percent_mito <span class="op">&lt;</span> <span class="dv">30</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="filter-out-genes-that-are-not-present-in-any-cells" class="level3">
<h3 class="anchored" data-anchor-id="filter-out-genes-that-are-not-present-in-any-cells">Filter out genes that are not present in any cells</h3>
<p>In general, we also want to get rid of genes with a low cell count to simplify our data.</p>
<div id="cell-49" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="d716827c-a4a6-4a8a-c42a-2f188c38783e">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_genes(adata, min_cells<span class="op">=</span><span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/preprocessing/_simple.py:293: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.
  adata.var["n_cells"] = number</code></pre>
</div>
</div>
</section>
<section id="visualizing-count-distributions" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-count-distributions">Visualizing count distributions</h3>
<p>Examination of the gene count and UMI count distributions is useful QC to evaluate the quality of the library and how deeply it was sequenced.</p>
<div id="cell-52" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:500}}" data-outputid="f47223c0-b6a4-48be-c1a1-25b026b0ea44">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'n_genes'</span>, <span class="st">'n_counts'</span>, <span class="st">'percent_mito'</span>], jitter<span class="op">=</span><span class="fl">0.4</span>, multi_panel<span class="op">=</span><span class="va">True</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-19-output-1.png" width="1505" height="483" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="step-4-normalization-and-variance-stabilization" class="level2">
<h2 class="anchored" data-anchor-id="step-4-normalization-and-variance-stabilization">Step 4: Normalization and Variance Stabilization</h2>
<p>In an scRNA-seq experiment, UMI counts provide a relative estimate of transcript abundance within each cell. During library preparation and sequencing, mRNA molecules from different cells are reverse transcribed and sequenced to varying degrees. As a result, UMI counts cannot be interpreted as absolute measurements of RNA abundance.</p>
<p>To make counts comparable across cells, we normalize the data so that each cell has a total of 1×10⁶ counts. This scaling preserves integer values while allowing gene expression levels to be expressed as relative abundances within each cell.</p>
<p>Downstream analyses such as PCA and differential expression typically assume that gene expression values are approximately continuous and have roughly constant variance across their range. However, raw scRNA-seq counts are <strong>overdispersed</strong> — they follow a negative binomial–like distribution, where the variance increases with the mean. To stabilize the variance and make the data more compatible with these methods, we apply a <strong>log transformation</strong> to the normalized counts.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This is a simplified approach for addressing overdispersion. Although log normalization is widely used, it does entirely produce constant-variance data. More sophisticated methods (e.g., variance-stabilizing transformations or model-based approaches) can better correct for overdispersion, but no perfect method currently exists — and those techniques are beyond the scope of this tutorial.</p>
</div>
</div>
<div id="cell-55" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata, target_sum<span class="op">=</span><span class="fl">1e6</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="cell-56" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="identify-highly-variable-genes" class="level3">
<h3 class="anchored" data-anchor-id="identify-highly-variable-genes">Identify Highly Variable Genes</h3>
<p>To simplify our data and reduce its dimensionality for downstream analyses, we limit our dataset to highly variable genes. Genes with low expression variance — such as housekeeping genes — contribute little to distinguishing between cell types or states, and can be excluded without loss of biological signal.</p>
<p>Here, we retain only genes with a mean expression between 0.01 and 8 (values above 8 likely indicate technical artifacts) and a minimum normalized dispersion of 1. Cells are grouped (binned) by mean gene expression, and genes within each bin are filtered based on their normalized dispersion values.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>sc.pp.highly_variable_genes</code> function in <strong>ScanPy</strong> expects input data that have already been <strong>normalized</strong> and <strong>variance-stabilized</strong>.</p>
</div>
</div>
<div id="cell-59" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:454}}" data-outputid="8a47116f-8ab3-4ac5-89bf-85becda1b2be">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, min_mean<span class="op">=</span><span class="fl">0.01</span>, max_mean<span class="op">=</span><span class="dv">8</span>, min_disp<span class="op">=</span><span class="dv">1</span>, n_bins<span class="op">=</span><span class="dv">20</span>, flavor<span class="op">=</span><span class="st">"seurat"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>sc.pl.highly_variable_genes(adata)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-22-output-1.png" width="1084" height="437" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Finally, we scale the the data to unit variance and zero mean to standardize gene expression by preventing highly expressed genes from dominating downstream analysis and ensuring all genes are weighted equally.</p>
<div id="cell-61" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="ef789c72-b270-4e33-9c9c-087afab057b7">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sc.pp.scale(adata, max_value<span class="op">=</span><span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/lib/python3.12/functools.py:912: UserWarning: zero-centering a sparse array/matrix densifies it.
  return dispatch(args[0].__class__)(*args, **kw)</code></pre>
</div>
</div>
</section>
</section>
<section id="step-5-clustering-and-visualization" class="level2">
<h2 class="anchored" data-anchor-id="step-5-clustering-and-visualization">Step 5: Clustering and Visualization</h2>
<p>When analyzing scRNA-seq data, <strong>clustering</strong> is used to identify distinct cell types and states. Subsequently, <strong>differential expression analysis</strong> can reveal the genes that define these cellular states.</p>
<p>There are many algorithms for clustering cells, and while they have been compared in detail in various benchmarks (see e.g., <a href="https://f1000research.com/articles/7-1141/v2">Duo et al.&nbsp;2018</a>), there is no univerally agreed upon method. Here we demonstrate clustering using <strong>Louvain clustering</strong>, which is a popular method for clustering single-cell RNA-seq data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Louvain clustering is a non-deterministic algorithm, meaning it can produce slightly different results across runs. The number and size of clusters depend on a resolution parameter, which controls the granularity of the clustering. There is no universally optimal way to choose this parameter — it typically depends on how specific or fine-grained you want your cell populations to be.</p>
</div>
</div>
<div id="cell-64" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="75be913b-5146-4204-86bf-29315c63184b">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sc.pp.pca(adata, mask_var<span class="op">=</span><span class="st">"highly_variable"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/preprocessing/_pca/__init__.py:245: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.
  Version(ad.__version__) &lt; Version("0.9")</code></pre>
</div>
</div>
<div id="cell-65" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="11f2503c-3daf-404a-b2eb-9d8fc13bd9ce">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster the cells using Louvain clustering</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_neighbors<span class="op">=</span><span class="dv">30</span>, n_pcs<span class="op">=</span><span class="dv">10</span>, knn<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>sc.tl.louvain(adata, resolution <span class="op">=</span> <span class="fl">1.0</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/neighbors/__init__.py:430: FutureWarning: Use obsm (e.g. `k in adata.obsm` or `adata.obsm.keys() | {'u'}`) instead of AnnData.obsm_keys, AnnData.obsm_keys is deprecated and will be removed in the future.
  if "X_diffmap" in adata.obsm_keys():</code></pre>
</div>
</div>
<p>It is useful to revisit the PCA projection with points colored by cluster. Previously we computed the PCA projection directly; here we use a function in ScanPy which does the same.</p>
<section id="visualization-with-pca" class="level3">
<h3 class="anchored" data-anchor-id="visualization-with-pca">Visualization with PCA</h3>
<div id="cell-68" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:621}}" data-outputid="b4b31b01-8b7c-4b28-b71b-3c781d86301d">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA and plot the projection to the first two dimensions, with points colored according to the Louvain clusters.</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">7</span>))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>sc.pl.pca(adata, color<span class="op">=</span><span class="st">'louvain'</span>, ax<span class="op">=</span>ax)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-26-output-1.png" width="896" height="604" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="nonlinear-visualization" class="level3">
<h3 class="anchored" data-anchor-id="nonlinear-visualization">Nonlinear Visualization</h3>
<p>The PCA representation results from a linear projection of the data from its original high-dimensional space to a lower-dimensional one (in this case, 2D). Such projections are valuable because they preserve global structure and are mathematically well-defined and reproducible.</p>
<p>In contrast, non-linear dimensionality reduction methods—the most popular being <strong>t-SNE</strong> and <strong>UMAP</strong>—can reveal complex, curved relationships in the data that linear methods cannot capture. However, they distort global distances and relationships, depend heavily on algorithmic parameters and random initialization, and may produce different results across runs. As a result, they are excellent for visualization but should be interpreted cautiously and not used for quantitative downstream analysis.</p>
<p>To read more, please see the <a href="http://biorxiv.org/lookup/doi/10.1101/2021.08.25.457696">Pachter Lab’s paper</a> on the drawbacks of non-linear dimensionality reduction methods.</p>
<section id="t-sne" class="level4">
<h4 class="anchored" data-anchor-id="t-sne">t-SNE</h4>
<div id="cell-71" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:621}}" data-outputid="39fc34a1-85d5-4ac9-a847-8c1f071f664b">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize cells with t-SNE. The n_pcs parameter sets the number of principal components to project to prior to</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="co"># performing t-SNE</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>sc.tl.tsne(adata, n_pcs<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">7</span>))</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>sc.pl.tsne(adata, color<span class="op">=</span><span class="st">'louvain'</span>, ax<span class="op">=</span>ax)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-27-output-1.png" width="896" height="604" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="umap" class="level4">
<h4 class="anchored" data-anchor-id="umap">UMAP</h4>
<div id="cell-73" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:657}}" data-outputid="ba593484-9fe1-44ca-a3d2-a426536b78e5">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>time</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">7</span>))</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span><span class="st">'louvain'</span>, ax<span class="op">=</span>ax)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-28-output-1.png" width="896" height="604" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>CPU times: user 3.07 s, sys: 7.18 ms, total: 3.08 s
Wall time: 2.97 s</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="step-6-differential-analysis" class="level2">
<h2 class="anchored" data-anchor-id="step-6-differential-analysis">Step 6: Differential Analysis</h2>
<p>From our clustering of cell states, we can perform <strong>differential expression analysis</strong> to identify <strong>marker genes</strong> that distinguish one cell state or cluster from another. This is typically done using simple statistical tests — such as the <strong>Wilcoxon rank-sum test</strong> or the <strong>t-test</strong> — which compare gene expression levels between clusters, or alternatively by fitting a more complex statistical model that accounts for variability across cells. By examining which genes are upregulated in each cluster, we can annotate clusters with known cell identities or infer novel cellular subpopulations.</p>
<p>Below, we use the <strong>Wilcoxon rank-sum test</strong> to identify marker genes. Because we are performing multiple hypothesis tests on the same dataset, we must adjust our p-values to control the false discovery rate. Several correction methods exist, but here we apply the <strong>Benjamini–Hochberg procedure</strong>, a widely used approach for multiple testing correction.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Statistical tests such as the <strong>Wilcoxon rank-sum test</strong> and the <strong>t-test</strong> assume that the groups being compared were defined independently of the data. However, in single-cell analysis, clustering algorithms like Louvain define groups based on the data itself — by minimizing within-cluster variation and maximizing between-cluster separation. This violates the independence assumption, meaning these tests are not strictly valid. Despite this, they remain common in practice and can still provide useful exploratory insights when interpreted cautiously.</p>
</div>
</div>
<div id="cell-76" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="1206dc4b-7a7f-4ceb-f9c3-43aa686c8f64">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, gene_symbols<span class="op">=</span><span class="st">'gene_names'</span>, groupby<span class="op">=</span><span class="st">'louvain'</span>, method<span class="op">=</span><span class="st">'wilcoxon'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(
/usr/local/lib/python3.12/dist-packages/scanpy/tools/_rank_genes_groups.py:482: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, "logfoldchanges"] = np.log2(</code></pre>
</div>
</div>
<p>The below plot shows the 10 most highly expressed genes in each cluster (compared to the other clusters) according to the Wilcoxon z-score.</p>
<div id="cell-78" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:1000}}" data-outputid="734e9fea-8282-4e27-9070-2293f7cb3004">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups(adata, n_genes<span class="op">=</span><span class="dv">10</span>, gene_symbols<span class="op">=</span><span class="st">'gene_names'</span>, groupby<span class="op">=</span><span class="st">'louvain'</span>, method<span class="op">=</span><span class="st">'wilcoxon'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="sample_pipeline_files/figure-html/cell-30-output-1.png" width="2057" height="1200" class="figure-img"></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><img src="black_Caltech_logo.png" alt="Caltech Logo" width="300"></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>